akka-remote/src/main/scala/akka/remote/serialization/ArteryControlMessageSerializer.scala

```diff
@@ -0,0 +1,213 @@
+/**
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.remote.serialization
+
+import akka.actor.{ ActorRef, Address, ExtendedActorSystem }
+import akka.protobuf.MessageLite
+import akka.remote.artery.OutboundHandshake.{ HandshakeReq, HandshakeRsp }
+import akka.remote.artery.compress.CompressionProtocol._
+import akka.remote.artery.compress.{ CompressionProtocol, CompressionTable }
+import akka.remote.artery.{ ActorSystemTerminating, ActorSystemTerminatingAck, Quarantined, SystemMessageDelivery }
+import akka.remote.{ ArteryControlFormats, MessageSerializer, UniqueAddress, WireFormats }
+import akka.serialization.{ BaseSerializer, Serialization, SerializationExtension, SerializerWithStringManifest }
+
+/** INTERNAL API */
+private[akka] object ArteryControlMessageSerializer {
+  private val QuarantinedManifest = "a"
+  private val ActorSystemTerminatingManifest = "b"
+  private val ActorSystemTerminatingAckManifest = "c"
+  private val HandshakeReqManifest = "d"
+  private val HandshakeRspManifest = "e"
+  private val ActorRefCompressionAdvertisementManifest = "f"
+  private val ActorRefCompressionAdvertisementAckManifest = "g"
+  private val ClassManifestCompressionAdvertisementManifest = "h"
+  private val ClassManifestCompressionAdvertisementAckManifest = "i"
+  private val SystemMessageEnvelopeManifest = "j"
+  private val SystemMessageDeliveryAckManifest = "k"
+  private val SystemMessageDeliveryNackManifest = "l"
+}
+
+/** INTERNAL API */
+private[akka] final class ArteryControlMessageSerializer(val system: ExtendedActorSystem) extends SerializerWithStringManifest with BaseSerializer {
+  import ArteryControlMessageSerializer._
+
+  private lazy val serialization = SerializationExtension(system)
+
+  override def manifest(o: AnyRef): String = o match { // most frequent ones first
+    case _: HandshakeReq ⇒ HandshakeReqManifest
+    case _: HandshakeRsp ⇒ HandshakeRspManifest
+    case _: SystemMessageDelivery.SystemMessageEnvelope ⇒ SystemMessageEnvelopeManifest
+    case _: SystemMessageDelivery.Ack ⇒ SystemMessageDeliveryAckManifest
+    case _: SystemMessageDelivery.Nack ⇒ SystemMessageDeliveryNackManifest
+    case _: Quarantined ⇒ QuarantinedManifest
+    case _: ActorSystemTerminating ⇒ ActorSystemTerminatingManifest
+    case _: ActorSystemTerminatingAck ⇒ ActorSystemTerminatingAckManifest
+    case _: CompressionProtocol.ActorRefCompressionAdvertisement ⇒ ActorRefCompressionAdvertisementManifest
+    case _: CompressionProtocol.ActorRefCompressionAdvertisementAck ⇒ ActorRefCompressionAdvertisementAckManifest
+    case _: CompressionProtocol.ClassManifestCompressionAdvertisement ⇒ ClassManifestCompressionAdvertisementManifest
+    case _: CompressionProtocol.ClassManifestCompressionAdvertisementAck ⇒ ClassManifestCompressionAdvertisementAckManifest
+    case _ ⇒
+      throw new IllegalArgumentException(s"Can't serialize object of type ${o.getClass} in [${getClass.getName}]")
+  }
```

Comment:
So, this will be a 12 case if-else on type, could it be more performant having an index or something to lookup the manifest instead?